SimG4CMS/Forward/test/SimG4FluxAnalyzer.cc

```diff
@@ -0,0 +1,167 @@
+  virtual void analyze(edm::Event const&, edm::EventSetup const&) override;
+
+  edm::Service<TFileService>                   fs_;
+  std::vector<std::string>                     lvNames_;
+  std::vector<edm::EDGetTokenT<ParticleFlux> > tok_PF_;
+  TTree                                       *tree_;
+
+  std::vector<std::string>                     detName_;
+  std::vector<int>                             detId_, pdgId_, vxType_;
+  std::vector<float>                           tof_, vtxX_, vtxY_, vtxZ_;
+  std::vector<float>                           hitPtX_, hitPtY_, hitPtZ_;
+  std::vector<float>                           momX_, momY_, momZ_;
+};
+
+SimG4FluxAnalyzer::SimG4FluxAnalyzer(const edm::ParameterSet& iConfig) {
+
+  usesResource("TFileService");
+
+  //now do whatever initialization is needed
+  lvNames_  = iConfig.getParameter<std::vector<std::string> >("LVNames");
+#ifdef EDM_ML_DEBUG
+  std::cout << "SimG4FluxAnalyzer:: for " << lvNames_.size() << " names:";
+  for (auto name : lvNames_) std::cout << " " << name;
+  std::cout << std::endl;
+#endif
+
+  for (auto name : lvNames_) {
+    std::string tagn = name+"ParticleFlux";
+    tok_PF_.push_back(consumes<ParticleFlux>(edm::InputTag("g4SimHits",tagn)));
+#ifdef EDM_ML_DEBUG
+    std::cout << "Flux source " << edm::InputTag("g4SimHits",tagn) <<std::endl;
+#endif
+  }
+}
+
+SimG4FluxAnalyzer::~SimG4FluxAnalyzer() {}
+
+void SimG4FluxAnalyzer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
+  std::vector<std::string> lvnames = {"TotemT1Part1","TotemT1Part2","TotemT1Part3","TotemT2Part1","TotemT2Part2","TotemT2Part3"};
+  desc.add<std::vector<std::string>>("LVNames",lvnames);
+  descriptions.add("SimG4FluxAnalyzer",desc);
+}
+
+void SimG4FluxAnalyzer::beginJob() {
+
+  tree_ = fs_->make<TTree>("Flux","ParticleFlux");
+  tree_->Branch("DetectorName", &detName_);
+  tree_->Branch("DetectorID",   &detId_);
+  tree_->Branch("ParticleCode", &pdgId_);
+  tree_->Branch("VertexType",   &vxType_);
+  tree_->Branch("ParticleTOF",  &tof_);
+  tree_->Branch("VertexX",      &vtxX_);
+  tree_->Branch("VertexY",      &vtxY_);
+  tree_->Branch("VertexZ",      &vtxZ_);
+  tree_->Branch("HitPointX",    &hitPtX_);
+  tree_->Branch("HitPointY",    &hitPtY_);
+  tree_->Branch("HitPointZ",    &hitPtZ_);
+  tree_->Branch("MomentumX",    &momX_);
+  tree_->Branch("MomentumY",    &momY_);
+  tree_->Branch("MomentumZ",    &momZ_);
+}
+
+void SimG4FluxAnalyzer::analyze(const edm::Event& iEvent, 
+				const edm::EventSetup& iSetup) {
+
+  //Loop over all flux containers
+  detName_.clear(); detId_.clear();  pdgId_.clear(); vxType_.clear();
+  tof_.clear();     vtxX_.clear();   vtxY_.clear();  vtxZ_.clear();
+  hitPtX_.clear();  hitPtY_.clear(); hitPtZ_.clear();
+  momX_.clear();    momY_.clear();   momZ_.clear();
+#ifdef EDM_ML_DEBUG
+  unsigned int k(0);
+#endif
+  for (auto token : tok_PF_) {
```

Comment:
@bsunanda - when you use an auto range for loop, please, consider the following:
```
T        // I'm copying this
T&       // I'm modifying this
const T& // I'm reading this
```
Those are your "defaults". When T is a fundamental type (built-in), though, you generally just revert to const T (no reference) for reading, because a copy is cheaper than aliasing.